### PR TITLE
Feature: multi pool router

### PR DESCRIPTION
**Pull request by danoctavian**
## Context

Add support for using multiple pools to allocate capacity for a cover.

Issue: https://github.com/NexusMutual/cover-router/issues/25

## Changes proposed in this pull request
Add a new function that takes as input the requested cover amount and the set of staking pools that can be used.

Use a greedy allocation approach to find the best pools to allocate to, keeping in mind, price does not follow a linear function past 90% capacity, so it can't be as simple as sorting pools by base price and allocate until you fill it up etc.

Full details on this approach are in the RFC:

https://www.notion.so/nxmcommunity/RFC-Multi-pool-cover-router-e4a1637910be49268cb52c5fb9fbe21d

## Test plan
Write unit tests for the computation and the parts dependant on it |(quoteEngine.js)

## Checklist

- [x] Rebased the base branch
- [x] Attached corresponding Github issue
- [x] Prefixed the name with the type of change (i.e. feat, chore, test)
- [ ] Performed a self-review of my own code
- [x] Followed the style guidelines of this project
- [x] Made corresponding changes to the documentation
- [x] Didn't generate new warnings
- [ ] Didn't generate failures on existing tests
- [x] Added tests that prove my fix is effective or that my feature works

## Review

When reviewing a PR, please indicate intention in comments using the following emojis:
* :cake: = Nice to have but not essential.
* :bulb: = Suggestion or a comment based on personal opinion
* :hammer: = I believe this should be changed.
* :thinking: = I don’t understand something, do you mind giving me more context?
* :rocket: = Feedback

### CODE REVIEW

**Comment by danoctavian** [comment]
WIP as some quoteEngine.js tests are still failing

**Comment by danoctavian** [comment]
fixed all tests except 1.

Fairly satisfied with the coverage but i want to think of more edge cases

File: test/unit/utils.js [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    const premium = useFixedPrice
+      ? calculateFixedPricePremiumPerYear(amountInWei, pool.basePrice)
+      : calculatePremiumPerYear(amountInWei, pool.basePrice, pool.initialCapacityUsed, pool.totalCapacity);
+
+    totalPremium = totalPremium.add(premium);
+  }
+
+  return totalPremium;
+};
+const calculateOptimalPoolAllocationBruteForce = (coverAmount, pools, useFixedPrice) => {
```
Comment (danoctavian) [inline test/unit/utils.js:57]: questionable if this should be in the code

I found it useful to check the result of the other function in complex questions without computing things by hand.

File: src/lib/premium-computations.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+ *
+ * To allocate the units U we take each unit at a time from 0 to U,
+ * and allocate each unit to the *best* pool available in terms of price
+ * as long as the pool has enough capacity for that unit.
+ *
+ * If there is no pool with capacity available for a unit of allocation the function returns
+ * with an empty list (aborts allocation completely - no partial allocations supported).
+ *
+ * UNIT_SIZE is dynamically computed as 0.1% (1 / UNIT_DIVISOR) of the cover amount.
+ * It has a minimum which is 1e18 (can't go below 1 ETH or 1 DAI).
```
Comment (shark0der) [inline src/lib/premium-computations.js:62]: 0.1% seems way too small, last we discussed we should split the cover in 10. Even if we split it in 10 parts - we should have a minimum, currently we're eyeballing it to be somewhere between 10k and 100k DAI (or equivalent). Needs a bit of research by going through covers and comparing gas costs and premiums since the gas cost should not make the cover overly expensive.

File: src/lib/premium-computations.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+  const amountInUnits = coverAmount.div(UNIT_SIZE).add(extra).toNumber();
+
+  // the amount padding is the amount added artificially added when adding a whole unit
+  // to account for the remainder
+  const amountPadding = extra === 1 ? UNIT_SIZE.sub(coverAmount.mod(UNIT_SIZE)) : BigNumber.from(0);
+
+  let lowestCost = BigNumber.from(0);
+  // Pool Id (number) -> Capacity Amount (BigNumber)
+  const lowestCostAllocation = {};
```
Comment (shark0der) [inline src/lib/premium-computations.js:86]: Based on the variable name `lowestCostAllocation` and the comment this seems to store the new allocations? `lowest` and `cost` seem redundant. Same for `lowestCost` - it seems to be just `premium`.

File: src/lib/premium-computations.js [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+  // the amount padding is subtracted from the last pool used so that all allocated amounts
+  // across pools sum up to coverAmount
+  lowestCostAllocation[lastPoolIdUsed] = lowestCostAllocation[lastPoolIdUsed].sub(amountPadding);
+
+  return { lowestCostAllocation, lowestCost };
+};
+const calculateOptimalPoolAllocation = (coverAmount, pools, useFixedPrice) => {
+  return calculateOptimalPoolAllocationGreedy(coverAmount, pools, useFixedPrice);
+};
```
Comment (shark0der) [inline src/lib/premium-computations.js:142]: :wastebasket:

File: src/lib/premium-computations.js (hunk 2 above)
Comment (danoctavian) [reply src/lib/premium-computations.js:62]: what's the downside to using say 1% instead of 10% if the computation still completes fast enough?

File: src/lib/premium-computations.js [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+      // we advance one unit size at a time
+      const amountInWei = UNIT_SIZE;
+
+      if (poolCapacityUsed[pool.poolId].add(amountInWei).gt(pool.totalCapacity)) {
+        // can't allocate unit to pool
+        continue;
+      }
+
+      const premium = useFixedPrice
+        ? calculateFixedPricePremiumPerYear(amountInWei, pool.basePrice)
```
[the remark this answers is not in the document]
Comment (danoctavian) [reply src/lib/premium-computations.js:109]: It's a good point 

The fixed price should be but i don't have any onchain guarantees that it is like IRL someone might set something higher or lower 

You could assert that going into it that's 1 option

Reasons why i didn't do this:

Why have a specialized function for this, if the same algorithm solves the problem for any cost function?

There's an optimization part, but IRL with a reasonable chunk size there's nothing to optimize - it runs fast enough. So i don't see why i'd add a specialized function.

What benefits do we have from adding a specialized function?

File: src/lib/premium-computations.js (hunk 3 above)
Comment (danoctavian) [reply src/lib/premium-computations.js]: agreed and addressed

File: src/lib/premium-computations.js [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        : calculatePremiumPerYear(amountInWei, pool.basePrice, poolCapacityUsed[pool.poolId], pool.totalCapacity);
+
+      if (premium.lt(lowestCostPerPool)) {
+        lowestCostPerPool = premium;
+        lowestCostPool = pool;
```
[the remark this answers is not in the document]
Comment (danoctavian) [reply src/lib/premium-computations.js]: fixed

File: src/lib/premium-computations.js (hunk 2 above)
Comment (shark0der) [reply src/lib/premium-computations.js]: Downside of using units that are too small is that we may end up allocating too many pools and the gas costs would offset the optimizations that we obtain from "efficiently" routing.

File: src/lib/premium-computations.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (shark0der) [reply src/lib/premium-computations.js]: Separation of concerns. Currently we may leave it as is - that's why I said "might deserve" and not "should have".

File: src/lib/premium-computations.js (hunk 2 above)
Comment (roxdanila) [reply src/lib/premium-computations.js]: 1 DAI is too low as a chunk size imo - imagine having a segment with 1 DAI in it. calibrating the right min amount depends on the gas costs, as Anatol suggested above. my gut feeling is that something like 5K - 10K DAI is more appropriate

File: src/lib/quoteEngine.js [hunk 7]
```diff
@@ -43,15 +48,101 @@ const calculatePremiumPerYear = (coverAmount, basePrice, initialCapacityUsed, to
...
+  const UNIT_SIZE = coverAmount.div(UNIT_DIVISOR).gt(MIN_UNIT_SIZE) ? coverAmount.div(UNIT_DIVISOR) : MIN_UNIT_SIZE;
+
+  // compute the extra amount of units (0 or 1) to be added based on the division remainder.
+  const extra = coverAmount.mod(UNIT_SIZE).gt(0) ? 1 : 0;
+
+  const amountInUnits = coverAmount.div(UNIT_SIZE).add(extra).toNumber();
+
+  // the amount padding is the amount added artificially added when adding a whole unit
+  // to account for the remainder
+  const amountPadding = extra === 1 ? UNIT_SIZE.sub(coverAmount.mod(UNIT_SIZE)) : BigNumber.from(0);
```
Comment (roxdanila) [inline src/lib/quoteEngine.js:89]: this doesn't look elegant at all. why do we need it? how can we implement this in a more elegant / cleaner way?

File: src/lib/quoteEngine.js [hunk 8]
```diff
@@ -43,15 +48,101 @@ const calculatePremiumPerYear = (coverAmount, basePrice, initialCapacityUsed, to
...
+    poolCapacityUsed[lowestCostPoolId] = poolCapacityUsed[lowestCostPoolId].add(UNIT_SIZE);
+
+    lastPoolIdUsed = lowestCostPoolId;
+  }
+
+  // the amount padding is subtracted from the last pool used so that all allocated amounts
+  // across pools sum up to coverAmount
+  allocations[lastPoolIdUsed] = allocations[lastPoolIdUsed].sub(amountPadding);
+
+  return { allocations, premium };
```
Comment (shark0der) [inline src/lib/quoteEngine.js:145]: You're not using the returned premium at all. Return the allocations directly.

File: src/lib/quoteEngine.js (hunk 7 above)
Comment (danoctavian) [reply src/lib/quoteEngine.js:89]: if you split the input amount in fixed size chunks to run the algorithm on there's a remainder that is not exactly equal to  one chunk.

The remainder still needs to be allocated somewhere so the way i solve this is I map it to one full chunk.

To me this is the logical solution of how to solve this issue, maybe the naming can be improved?

File: src/lib/quoteEngine.js [hunk 9]
```diff
@@ -43,15 +48,101 @@ const calculatePremiumPerYear = (coverAmount, basePrice, initialCapacityUsed, to
...
+  }
 
-  return basePremium.add(surgePremium);
+  let lastPoolIdUsed;
+  for (let i = 0; i < amountInUnits; i++) {
+    let lowestCostPerPool = MaxUint256;
+    let lowestCostPoolId;
+    for (const pool of pools) {
+      // we advance one unit size at a time
+      const amountInWei = UNIT_SIZE;
```
Comment (shark0der) [inline src/lib/quoteEngine.js:108]: `amountInWei` variable is pointless - just use `UNIT_SIZE`. Bonus points if you rename `UNIT_SIZE` to `unitSize`

File: src/lib/quoteEngine.js (hunk 7 above)
Comment (danoctavian) [reply src/lib/quoteEngine.js]: eliminated using @shark0der  suggestion of substracting instead of running a for loop from 0 to N

[the remark this answers is not in the document]
Comment (danoctavian) [reply src/lib/quoteEngine.js]: great idea, used this

File: src/lib/premium-computations.js (hunk 2 above)
Comment (danoctavian) [reply src/lib/premium-computations.js]: adressed using 10%

File: src/lib/quoteEngine.js [hunk 10]
```diff
@@ -78,79 +157,88 @@ const quoteEngine = (store, productId, amount, period, coverAsset) => {
...
+      poolId,
+      basePrice,
+      initialCapacityUsed,
+      totalCapacity,
+    };
+  });
+
+  const { assets } = store.getState();
+  const daiRate = assetRates[assets.DAI];
+  const minUnitSizeInNxm = MIN_UNIT_SIZE_DAI.mul(WeiPerEther).div(daiRate);
```
Comment (danoctavian) [inline src/lib/quoteEngine.js:187]: note we are adding dynamic computation of unit size in NXM based on the fixed value of 10.000 DAI

**Review by shark0der — APPROVED** [review]
